File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.py
class Solution:
    def setZeroes(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        # Approach1: O(m x n) and O(m + n)
        m = len(matrix) # no. of rows
        n = len(matrix[0]) # no. of columns

        rows = set()
        cols = set()

        for i in range(m):
            for j in range(n):
                if matrix[i][j] == 0:
                    rows.add(i)
                    cols.add(j)

        for i in range(m):
            for j in range(n):
                if i in rows or j in cols:
                    matrix[i][j] = 0

        # Approach 2: O(m x n) and O(1)
        # In this approach, instead of using O(mxn) space complexity, there is another solution, where we can use
        # first row and first colunn of an array as storage. 
        

                

        

        
        

File: 0073-set-matrix-zeroes/0073-set-matrix-zeroes.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var setZeroes = function(matrix) {
    
    let m = matrix.length  // no. of rows
    let n = matrix[0].length // no. of cols

    let rows = new Set() // set to store the rows where zero element is present
    let cols = new Set() // set to store the cols where zero element is present

    // finding the rows and cols that needs to be zeroed
    for (let i = 0; i < m; i++){
        for (let j = 0; j < n; j++){
            if (matrix[i][j] == 0){
                rows.add(i)
                cols.add(j)
            }
        }
    }

    // making the elements from the indices from rows and cols to zero
    for (let i = 0; i < m; i++){
        for (let j = 0; j < n; j++){
            if (rows.has(i) || cols.has(j)){
                matrix[i][j] = 0
            }
        }
    }
    
};